content/post/postgres/rfc-extension-packaging-lookup.md

```diff
@@ -0,0 +1,496 @@
+inspired this proposal. These threads cover quite a lot of territory, so I
+thought it would be useful to pull together a more unified, public proposal.
+
+## The Problem
+
+A number of challenges face extension users in various configurations, thanks
+to extension file organization in the Postgres core. The common thread among
+them is the need to add extensions without changing the Postgres installation
+itself.
+
+### Packager Testing
+
+On Debian systems, the user account that creates extension packages lacks
+permission to add files to Postgres install. But testing extensions requires
+installing the extension files where Postgres can find them. Furthermore,
+extensions should ideally be built against a clean Postgres install; adding an
+extension in order to run `make installcheck` would pollute it.
+
+[Christoph's patch][destdir] solves these problems by adding a second lookup
+path for extensions and dynamic modules, so that Postgres can load them
+directly from the package build directory.
+
+Alas, the patch isn't ideal, because it simply specifies a prefix and appends
+the full `pg_config` directory paths to it. For example, if `--sharedir`
+outputs `/opt/share` and `extension_destdir` GUC is set to `/tmp/build/myext`,
+the patch will search in `/tmp/build/myext/opt/share`. This approach works for
+the packaging use case, which explicitly uses full paths with a prefix, but
+would be weird for other use cases.
+
+### Docker Immutability
+
+Docker images are immutable. To install persistent extensions in a running
+Docker container, one must create a persistent volume, map it to
+`SHAREDIR/extensions`, and copy over all the core extensions (or muck with
+[symlink magic]). Then do it again for shared object libraries (`PKGLIBDIR`),
+and perhaps also for other `pg_config` directories, like `--bindir`. Once it's
+all set up, one can install a new extension and its files will be distributed
+to the relevant persistent volumes.
+
+This pattern makes upgrades tricky, because the core extensions are mixed in
+with third-party extensions. Worse, the number of directories that must be
+mounted into volumes depends on the features of an extension, increasing
+deployment configuration complexity. It would be preferable to have all the
+files for an extension in one place, rather than scattered across multiple
+persistent volumes.
+
+### Postgres.app Immutability
+
+The macOS [Postgres.app] supports extensions. But installing one into
+`SHAREDIR/extensions` changes the contents of the Postgres.app bundle,
+breaking Apple-required signature validation. The OS will no longer be able to
+validate that the app is legit and refuse to start it.
+
+## Solution
+
+To address these issues, this RFC proposes to change file organization and
+lookup patterns for PostgreSQL extensions.
+
+### Extension Directories
+
+First, when an extension is installed, all of its files will live in a single
+directory named for the extension. The contents include:
```

Comment:
This is not a new problem as of today, but I believe this RFC would be the place to solve it: how do we ensure extension names --and hence directory names-- are unique? TBH I don't have a good answer today, but I believe this should be discussed and addressed given the potential window of opportunity of changes in the extension system at the Postgres core.

Quick initial thoughts, probably not good ideas anyway: since there's no "central authority", possibly the only reasonable option is to use a reverse DNS name (e.g. `com.vendor.ourextension`) or, to make it easier, something like `$extname-$fqdn`. The `create extension` command could default to try to find by name only, and if there are no ambiguities, proceed without requiring the user to fully qualify it. The same would apply to the steps followed by Postgres to load the objects from the extensions.

Reply:
I like the idea of a DNS-based directory format; I went on at length about it [back in February](https://justatheory.com/2024/02/decentralized-extension-publishing/). I decided it could be addressed later, or at least outside this proposal, since it has implications for more SQL syntax to be able to load one `foo` extension vs. another. Probably C namespacing, too.

Reply:
I agree it could be part of another proposal, since it's slightly tangential. However, I'd consider then proposing both RFCs together, as any change in core should in my opinion address "all the relevant issues as of today", or else we will end up with a half-baked solution that won't be touched again in, possibly, years. Let's use the momentum we may have to propose changes to consider all that we may need.

Reply:
I'm in favor, but I think someone with much more experience with C extensions, compiler options, and whatnot needs to work on it. Any suggestions?

Reply:
LOL, I've generally had more luck working directly with an interested person or likely candidate than soliciting volunteers. Or just doing it myself, which I'd eventually be willing to do. But it will be some time before I have that bandwidth.